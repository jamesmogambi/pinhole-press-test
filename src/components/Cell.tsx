import React from "react";
import cellIcon from "../icon.svg";

type CellProp = {
  value: string | number;
  icon: boolean;
  styleProp: any;
};

const Cell: React.FC<CellProp> = ({ value, styleProp, icon }) => {
  return (
    <div className={`${styleProp} cell`}>
      {icon && <img src={cellIcon} alt="icon" className="icon" />}
      {value}
    </div>
  );
};

export default Cell;
