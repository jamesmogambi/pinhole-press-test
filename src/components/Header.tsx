import React from "react";

const Header: React.FC<{ states: IState[] }> = ({ states }) => {
  return (
    <React.Fragment>
      <div className="header">
        <div className="container shadow">
          <span className="title">PRODUCTS</span>
          <div className="items">
            <span className="item">Category</span>
            <span className="item">Sub-Category</span>
          </div>
        </div>
        <div className="container">
          <span className="title title-state">STATE</span>
          <div className="items">
            {states.map((s, i) => (
              <span key={i} className="item">
                {s.state}
              </span>
            ))}
          </div>
        </div>
      </div>
    </React.Fragment>
  );
};

export default Header;
