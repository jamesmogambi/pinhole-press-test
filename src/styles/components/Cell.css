.cell {
  width: 120px;
  font-size: 15px;
  padding: 10px;
  color: rgb(78, 78, 78);
  margin: 0px;
}

.cell-dark {
  background-color: var(--dark-gray-color);
}

.cell-light {
  background-color: var(--white-color);
  text-align: center;
}

.text-dark {
  color: #000000;
  font-weight: bold;
}

.text-light {
  color: var(--dark-gray-color);
}

.cell-double {
  width: 240px;
}

.cell-total {
  text-align: center;
}

.icon {
  height: 20px;
  margin-right: 10px;
}

.cell-icon {
  display: flex;
}
