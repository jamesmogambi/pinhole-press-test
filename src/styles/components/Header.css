.header {
  background: var(--dark-blue-color);
  height: 105px;
  display: flex;
  padding-top: 7px;
}

.title {
  color: var(--white-color);
  font-weight: bold;
  padding-left: 10px;
}

.item {
  color: rgba(250, 235, 215, 0.774);
  width: 120px;
  font-size: 15px;
  padding: 10px;
}

.items {
  margin-top: 20px;
  display: flex;
  flex-direction: row;
}

.title-state {
  position: relative;
  left: 8%;
}
