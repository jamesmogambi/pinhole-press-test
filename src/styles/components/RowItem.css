.row-items,
.column {
  display: flex;
}

.row-items {
  flex-direction: row;
}

.column {
  flex-direction: column;
}
