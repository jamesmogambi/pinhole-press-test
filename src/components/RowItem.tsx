import React from "react";
import { connect } from "react-redux";
import {
  getSubCategories,
  getSubTotal,
  getTotal,
} from "../selectors/categories";
import Cell from "./Cell";

interface RowItemProp {
  category: Category;
  states: IState[];
  sales: ISale[];
  subcategories: SubCategory[];
}

export const RowItem: React.FC<RowItemProp> = (props) => {
  let { category, subcategories, states, sales } = props;

  const renderSubTotal = (item: SubCategory) => {
    return states.map((s, i) => {
      let { state } = s;
      let subTotal = getSubTotal(item, state, sales);
      return (
        <Cell styleProp="cell-light" key={i} value={subTotal} icon={false} />
      );
    });
  };

  const renderTotal = () => {
    return states.map((s, i) => {
      let { state } = s;
      let total = getTotal(state, category, sales);
      return (
        <Cell
          key={i}
          styleProp="text-dark cell-dark cell-total"
          value={total}
          icon={false}
        />
      );
    });
  };
  return (
    <div>
      <div className="row-items">
        <Cell
          value={category.name}
          styleProp="text-dark cell-icon"
          icon={true}
        />
        <div className="column">
          {subcategories.map((item, i) => (
            <div key={i} className="row-items">
              <Cell styleProp={null} value={item.subCategory} icon={false} />
              <div className="row-items">{renderSubTotal(item)}</div>
            </div>
          ))}
        </div>
      </div>
      <div className="row-items">
        <Cell
          styleProp="text-dark cell-double"
          value={`${category.name} total`}
          icon={false}
        />
        <div className="row-items">{renderTotal()}</div>
      </div>
    </div>
  );
};

const mapStateToProps = (state: any, ownProps: any) => {
  let { sales } = state.sale;
  let { category } = ownProps;
  return {
    sales,
    subcategories: getSubCategories(category, sales),
  };
};

export default connect(mapStateToProps)(RowItem);
