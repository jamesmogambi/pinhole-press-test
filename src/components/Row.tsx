import React from "react";
import RowItem from "./RowItem";
import Cell from "./Cell";
import { getGrandTotal } from "../selectors/categories";

interface RowProp {
  states: IState[];
  sales: ISale[];
  categories: Category[];
}

export const Row: React.FC<RowProp> = ({ states, categories, sales }) => {
  const renderGrandTotal = () => {
    return states.map((s, i) => {
      let { state } = s;
      let grandTotal = getGrandTotal(state, sales);
      return (
        <Cell
          key={i}
          styleProp="text-dark cell-dark cell-total"
          value={grandTotal}
          icon={false}
        />
      );
    });
  };

  return (
    <React.Fragment>
      <div className="container">
        {categories.map((item, i) => (
          <RowItem key={i} category={item} states={states} />
        ))}
        <div className="row-items item-border">
          <Cell
            styleProp="text-dark cell-double"
            value="Grand Total"
            icon={false}
          />
          {renderGrandTotal()}
        </div>
      </div>
    </React.Fragment>
  );
};

export default Row;
