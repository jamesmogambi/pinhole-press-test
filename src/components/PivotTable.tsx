import React from "react";
import { connect } from "react-redux";
import selectStates from "../selectors/states";
import { getCategories } from "../selectors/categories";
import Header from "./Header";
import Row from "./Row";

interface PivotTableProps {
  sales: ISale[];
  states: IState[];
  categories: Category[];
}
export const PivotTable: React.FC<PivotTableProps> = (props) => {
  return (
    <React.Fragment>
      <div className="scroll bg">
        <Header states={props.states} />
        <Row
          states={props.states}
          categories={props.categories}
          sales={props.sales}
        />
      </div>
    </React.Fragment>
  );
};

const mapStateToProps = (state: any) => {
  let { sales } = state.sale;
  return {
    states: selectStates(sales),
    sales,
    categories: getCategories(sales),
  };
};

export default connect(mapStateToProps)(PivotTable);
