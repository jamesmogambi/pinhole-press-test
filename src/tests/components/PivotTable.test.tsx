import { shallow } from "enzyme";
import { PivotTable } from "../../components/PivotTable";
import sales from "../fixtures/sales";

test("should correctly render Row with props", () => {
  const categories = [{ name: "Furniture" }, { name: "Technology" }];
  const states = [
    { state: "Alabama", city: "Decatur" },
    { state: "Arizona", city: "Gilbert" },
    { state: "Arkansas", city: "Fayetteville" },
  ];

  const wrapper = shallow(
    <PivotTable sales={sales} states={states} categories={categories} />
  );
  expect(wrapper).toMatchSnapshot();
});
