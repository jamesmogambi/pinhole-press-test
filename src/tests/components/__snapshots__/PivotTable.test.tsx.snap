// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Row with props 1`] = `
<Fragment>
  <div
    className="scroll bg"
  >
    <Header
      states={
        Array [
          Object {
            "city": "Decatur",
            "state": "Alabama",
          },
          Object {
            "city": "Gilbert",
            "state": "Arizona",
          },
          Object {
            "city": "Fayetteville",
            "state": "Arkansas",
          },
        ]
      }
    />
    <Row
      categories={
        Array [
          Object {
            "name": "Furniture",
          },
          Object {
            "name": "Technology",
          },
        ]
      }
      sales={
        Array [
          Object {
            "category": "Furniture",
            "city": "Decatur",
            "country": "United States",
            "customerId": "CG-12520",
            "customerName": "Claire Gute",
            "discount": 0,
            "orderDate": "11/8/16",
            "orderId": "CA-2016-152156",
            "postalCode": 42420,
            "productId": "FUR-BO-10001798",
            "productName": "Bush Somerset Collection Bookcase",
            "profit": 41.9136,
            "quantity": 2,
            "region": "South",
            "rowId": 1,
            "sales": 261.96,
            "segment": "Consumer",
            "shipDate": "11/11/16",
            "shipMode": "Second Class",
            "state": "Alabama",
            "subCategory": "Bookcases",
          },
          Object {
            "category": "Furniture",
            "city": "Decatur",
            "country": "United States",
            "customerId": "CG-12520",
            "customerName": "Claire Gute",
            "discount": 0,
            "orderDate": "11/8/16",
            "orderId": "CA-2016-152156",
            "postalCode": 42420,
            "productId": "FUR-CH-10000454",
            "productName": "Hon Deluxe Fabric Upholstered Stacking Chairs, Rounded Back",
            "profit": 219.582,
            "quantity": 3,
            "region": "South",
            "rowId": 2,
            "sales": 731.94,
            "segment": "Consumer",
            "shipDate": "11/11/16",
            "shipMode": "Second Class",
            "state": "Alabama",
            "subCategory": "Bookcases",
          },
          Object {
            "category": "Furniture",
            "city": "Gilbert",
            "country": "United States",
            "customerId": "DV-13045",
            "customerName": "Darrin Van Huff",
            "discount": 0,
            "orderDate": "6/12/16",
            "orderId": "CA-2016-138688",
            "postalCode": 90036,
            "productId": "OFF-LA-10000240",
            "productName": "Self-Adhesive Address Labels for Typewriters by Universal",
            "profit": 6.8714,
            "quantity": 2,
            "region": "West",
            "rowId": 3,
            "sales": 14.62,
            "segment": "Corporate",
            "shipDate": "6/16/16",
            "shipMode": "Second Class",
            "state": "Arizona",
            "subCategory": "Chairs",
          },
          Object {
            "category": "Technology",
            "city": "Los Angeles",
            "country": "United States",
            "customerId": "SO-20335",
            "customerName": "Sean O'Donnell",
            "discount": 0.45,
            "orderDate": "10/11/15",
            "orderId": "US-2015-108966",
            "postalCode": 33311,
            "productId": "FUR-TA-10000577",
            "productName": "Bretford CR4500 Series Slim Rectangular Table",
            "profit": -383.031,
            "quantity": 5,
            "region": "South",
            "rowId": 4,
            "sales": 957.5775,
            "segment": "Consumer",
            "shipDate": "10/18/15",
            "shipMode": "Standard Class",
            "state": "California",
            "subCategory": "Accessories",
          },
          Object {
            "category": "Technology",
            "city": "Gilbert",
            "country": "United States",
            "customerId": "SO-20335",
            "customerName": "Sean O'Donnell",
            "discount": 0.45,
            "orderDate": "10/11/15",
            "orderId": "US-2015-108966",
            "postalCode": 33311,
            "productId": "FUR-TA-10000577",
            "productName": "Bretford CR4500 Series Slim Rectangular Table",
            "profit": -383.031,
            "quantity": 5,
            "region": "South",
            "rowId": 5,
            "sales": 556.3452,
            "segment": "Consumer",
            "shipDate": "10/18/15",
            "shipMode": "Standard Class",
            "state": "Arizona",
            "subCategory": "Accessories",
          },
          Object {
            "category": "Furniture",
            "city": "Decatur",
            "country": "United States",
            "customerId": "CG-12520",
            "customerName": "Claire Gute",
            "discount": 0,
            "orderDate": "11/8/16",
            "orderId": "CA-2016-152156",
            "postalCode": 42420,
            "productId": "FUR-BO-10001798",
            "productName": "Bush Somerset Collection Bookcase",
            "profit": 41.9136,
            "quantity": 2,
            "region": "South",
            "rowId": 6,
            "sales": 700.33,
            "segment": "Consumer",
            "shipDate": "11/11/16",
            "shipMode": "Second Class",
            "state": "Alabama",
            "subCategory": "Bookcases",
          },
        ]
      }
      states={
        Array [
          Object {
            "city": "Decatur",
            "state": "Alabama",
          },
          Object {
            "city": "Gilbert",
            "state": "Arizona",
          },
          Object {
            "city": "Fayetteville",
            "state": "Arkansas",
          },
        ]
      }
    />
  </div>
</Fragment>
`;
